### PR TITLE
DSD-1797: Breadcrumbs custom link prop

### PR DESCRIPTION
**Pull request by 7emansell**
Fixes JIRA ticket [DSD-1797](https://newyorkpubliclibrary.atlassian.net/browse/DSD-1797).

## This PR does the following:

- Adds `customLinkComponent` prop to `Breadcrumbs`
- Adds `linkProps` prop to `breadcrumbsData`

## How has this been tested?

Locally, and in [this Turbine branch](https://nypl-ds-test-b59lbu4si-nypl.vercel.app/components/breadcrumbs)

### Accessibility Checklist

- [x] Checked Storybook's "Accessibility" tab for color contrast and other issues.
- [x] The feature works with keyboard inputs including tabbing back and forward and pressing space, enter, arrow, and esc keys.
- [x] For hidden text or when `aria-live` is used, a screenreader was used to verify the text is read.
- [x] For features that involve UI updates and focusing on DOM `ref`s, focus management was reviewed.
- [x] The feature works when the page is zoomed in to 200% and 400%.

## Open Questions

- `customLink` is resistant to stricter typing but I would like for it not to be `any`?

### Checklist:

- [x] I have updated the Storybook documentation and changelog accordingly.
- [x] I have added relevant accessibility documentation for this pull request.
- [x] All new and existing tests passed.

### Front End Review:

- [x] Review the Vercel preview deployment once it is ready.

[DSD-1797]: https://newyorkpubliclibrary.atlassian.net/browse/DSD-1797?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/components/Breadcrumbs/Breadcrumbs.mdx [hunk 1]
```diff
@@ -69,6 +70,31 @@ the truncated page title.
 
 <Canvas of={BreadcrumbsStories.LongText} />
 
+## Custom Link and Link Props
+
+The `customLink` prop allows apps with internal routers (Next.js, for example) to pass their custom link component into Breadcrumbs so the links can use client-side routing.
```
Comment (EdwinGuzman) [inline src/components/Breadcrumbs/Breadcrumbs.mdx]: nit: "into the `Breadcrumbs` component so ...."

File: src/components/Breadcrumbs/Breadcrumbs.mdx [hunk 2]
```diff
@@ -69,6 +70,31 @@ the truncated page title.
...
+To maximize these benefits, the `linkProps` prop on `breadcrumbsData` allows accompanying custom props on a link-by-link basis.
```
Comment (EdwinGuzman) [inline src/components/Breadcrumbs/Breadcrumbs.mdx]: > To maximize these benefits,

I think I know what you're trying to say but you mentioned client-side routing and I'm assuming that's a benefit. I think you can add one or two more benefits or go into why client-side routing is beneficial in an app before this sentence.

File: src/components/Breadcrumbs/Breadcrumbs.mdx [hunk 3]
```diff
@@ -69,6 +70,31 @@ the truncated page title.
...
+
+Note: assigning a value for `href` in `linkProps` will override the `url` prop, so avoid this.
```
Comment (EdwinGuzman) [inline src/components/Breadcrumbs/Breadcrumbs.mdx]: We try to include "recommend" in these types of notes so how about, ", ... we strongly recommend against this." or something along those lines.

File: src/components/Breadcrumbs/Breadcrumbs.mdx [hunk 4]
```diff
@@ -69,6 +70,31 @@ the truncated page title.
...
+<Source
+  code={`
+  // Example with Next.js Link and associated props.
+const breadcrumbsData = [
+  { url: "#", text: "Parent", linkProps: { replace: true, scroll: false }}
+  { url: "#", text: "Child", , linkProps: { replace: true }}
+  { url: "#", text: "Grandchild" },
+];
+// ...
+<Breadcrumbs customLink={Link} breadcrumbsData={breadcrumbsData} />
```
Comment (EdwinGuzman) [inline src/components/Breadcrumbs/Breadcrumbs.mdx]: At the top of this code example snippet, add the import for the `Link` component from next/link, to explicitly show where it's imported from.

File: src/components/Breadcrumbs/Breadcrumbs.mdx (hunk 2 above)
Comment (EdwinGuzman) [reply src/components/Breadcrumbs/Breadcrumbs.mdx]: > accompanying custom props on a link-by-link basis

Can you also describe some examples and link to the nextjs link prop docs? You show it below in the code snippet but it'll be useful to say that it's really any prop that the framework's link component accepts.

File: src/components/Breadcrumbs/Breadcrumbs.tsx [hunk 5]
```diff
@@ -1,10 +1,10 @@
 import {
   Breadcrumb as ChakraBreadcrumb,
   BreadcrumbItem,
-  BreadcrumbLink,
   chakra,
   useStyleConfig,
   ChakraComponent,
+  BreadcrumbLink,
```
Comment (EdwinGuzman) [inline src/components/Breadcrumbs/Breadcrumbs.tsx]: Why the move?

File: src/components/Breadcrumbs/breadcrumbsChangelogData.ts [hunk 6]
```diff
@@ -9,6 +9,13 @@
 import { ChangelogData } from "../../utils/ComponentChangelogTable";
 
 export const changelogData: ChangelogData[] = [
+  {
+    date: "Prerelease",
+    version: "Prerelease",
+    type: "Update",
+    affects: ["Functionality"],
+    notes: ["Adds `customLink` and `linkProps` props."],
```
Comment (EdwinGuzman) [inline src/components/Breadcrumbs/breadcrumbsChangelogData.ts]: Add why they're added or what it supports.

**Review by EdwinGuzman — CHANGES_REQUESTED** [review]
Looks good but some documentation update. I haven't tested it beyond that time we tested it together and it was working but I'll try to review it in Turbine or DCF if you have a working branch/vercel preview.

**Comment by 7emansell** [comment]
> Looks good but some documentation update. I haven't tested it beyond that time we tested it together and it was working but I'll try to review it in Turbine or DCF if you have a working branch/vercel preview.

https://nypl-ds-test-b59lbu4si-nypl.vercel.app/components/breadcrumbs!

**Review by EdwinGuzman — APPROVED** [review]
Looks and works great. I'm debating whether `as` is a better prop name than `customLink`. `as` follows Chakra's convention and that's consistent.

There are various components that use the DS Link internally but I think only `Heading` and `Pagination` would benefit from using the nextjs link (from a quick glance). So if you were to add this feature to those components, would `as` or `customLink` work better?

I'm approving the functionality but updating the name is a small update.

**Comment by 7emansell** [comment]
I feel like `customLink` is clearer to me as a dev than `as` (which even as a Chakra convention is slightly confusing) but I agree either could work, so happy to change it!

**Review by jackiequach — APPROVED** [review]
I think that the `customLink` prop makes sense here since `as` would imply the `Breadcrumbs` component is a `Link`, which is not true since it would be the children of the component would be links.

**Comment by bigfishdesign13** [comment]
@7emansell,
I don't think `as`, in this case, is appropriate.  As Jackie noted, it doesn't feel right as a component level prop name.  Something like `linkAs` would be better.  That said, I am leaning toward the other naming convention.  But I think even that could be better.  `customLinkType` or `customLinkComponent` feels better to me.

Cc @jackiequach @EdwinGuzman

**Pull request merged**